File: greenlang/factors/sdk/python/pagination.py
# -*- coding: utf-8 -*-
"""Cursor + offset paginators for list/search responses.

Usage (sync)::

    for factor in client.paginate_search("diesel", page_size=50):
        print(factor.factor_id)

Usage (async)::

    async for factor in aclient.paginate_search("diesel", page_size=50):
        print(factor.factor_id)
"""
from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import (
    Any,
    AsyncIterator,
    Awaitable,
    Callable,
    Dict,
    Iterator,
    List,
    Optional,
    TypeVar,
)

logger = logging.getLogger(__name__)

T = TypeVar("T")


@dataclass
class PageInfo:
    """Summary of pagination progress."""

    page: int
    page_size: int
    total_count: Optional[int]
    next_cursor: Optional[str] = None
    items_yielded: int = 0


class OffsetPaginator(Iterator[T]):
    """Generic offset/limit iterator.

    The fetcher callable receives ``(offset, limit)`` and must return
    ``(items, total_count_or_none)``.  Iteration stops when:

        * ``items`` is empty, OR
        * ``offset + len(items) >= total_count`` (if total known), OR
        * user-supplied ``max_items`` was reached.
    """

    def __init__(
        self,
        fetcher: Callable[[int, int], "tuple[List[T], Optional[int]]"],
        *,
        page_size: int = 100,
        start_offset: int = 0,
        max_items: Optional[int] = None,
    ) -> None:
        self._fetcher = fetcher
        self._page_size = max(1, int(page_size))
        self._offset = int(start_offset)
        self._max_items = max_items
        self._buffer: List[T] = []
        self._total: Optional[int] = None
        self._exhausted = False
        self._yielded = 0

    def __iter__(self) -> Iterator[T]:
        return self

    def __next__(self) -> T:
        if self._max_items is not None and self._yielded >= self._max_items:
            raise StopIteration
        if not self._buffer and not self._exhausted:
            self._fill()
        if not self._buffer:
            raise StopIteration
        item = self._buffer.pop(0)
        self._yielded += 1
        return item

    def _fill(self) -> None:
        limit = self._page_size
        if self._max_items is not None:
            limit = min(limit, self._max_items - self._yielded)
            if limit <= 0:
                self._exhausted = True
                return
        items, total = self._fetcher(self._offset, limit)
        self._buffer.extend(items)
        self._offset += len(items)
        if total is not None:
            self._total = total
        if not items or (total is not None and self._offset >= total):
            self._exhausted = True

    @property
    def total_count(self) -> Optional[int]:
        return self._total


class CursorPaginator(Iterator[T]):
    """Generic cursor-based iterator.

    The fetcher callable receives ``cursor`` (None on first call) and
    must return ``(items, next_cursor)``.  Iteration stops when
    ``next_cursor`` is ``None`` or user-supplied ``max_items`` reached.
    """

    def __init__(
        self,
        fetcher: Callable[[Optional[str]], "tuple[List[T], Optional[str]]"],
        *,
        max_items: Optional[int] = None,
    ) -> None:
        self._fetcher = fetcher
        self._max_items = max_items
        self._buffer: List[T] = []
        self._cursor: Optional[str] = None
        self._started = False
        self._exhausted = False
        self._yielded = 0

    def __iter__(self) -> Iterator[T]:
        return self

    def __next__(self) -> T:
        if self._max_items is not None and self._yielded >= self._max_items:
            raise StopIteration
        if not self._buffer and not self._exhausted:
            self._fill()
        if not self._buffer:
            raise StopIteration
        item = self._buffer.pop(0)
        self._yielded += 1
        return item

    def _fill(self) -> None:
        if self._started and self._cursor is None:
            self._exhausted = True
            return
        items, next_cursor = self._fetcher(self._cursor)
        self._started = True
        self._buffer.extend(items)
        self._cursor = next_cursor
        if not items or next_cursor is None:
            self._exhausted = True

    @property
    def cursor(self) -> Optional[str]:
        return self._cursor


class AsyncOffsetPaginator:
    """Async sibling of :class:`OffsetPaginator`."""

    def __init__(
        self,
        fetcher: Callable[[int, int], Awaitable["tuple[List[T], Optional[int]]"]],
        *,
        page_size: int = 100,
        start_offset: int = 0,
        max_items: Optional[int] = None,
    ) -> None:
        self._fetcher = fetcher
        self._page_size = max(1, int(page_size))
        self._offset = int(start_offset)
        self._max_items = max_items
        self._buffer: List[T] = []
        self._total: Optional[int] = None
        self._exhausted = False
        self._yielded = 0

    def __aiter__(self) -> AsyncIterator[T]:
        return self

    async def __anext__(self) -> T:
        if self._max_items is not None and self._yielded >= self._max_items:
            raise StopAsyncIteration
        if not self._buffer and not self._exhausted:
            await self._fill()
        if not self._buffer:
            raise StopAsyncIteration
        item = self._buffer.pop(0)
        self._yielded += 1
        return item

    async def _fill(self) -> None:
        limit = self._page_size
        if self._max_items is not None:
            limit = min(limit, self._max_items - self._yielded)
            if limit <= 0:
                self._exhausted = True
                return
        items, total = await self._fetcher(self._offset, limit)
        self._buffer.extend(items)
        self._offset += len(items)
        if total is not None:
            self._total = total
        if not items or (total is not None and self._offset >= total):
            self._exhausted = True


class AsyncCursorPaginator:
    """Async sibling of :class:`CursorPaginator`."""

    def __init__(
        self,
        fetcher: Callable[[Optional[str]], Awaitable["tuple[List[T], Optional[str]]"]],
        *,
        max_items: Optional[int] = None,
    ) -> None:
        self._fetcher = fetcher
        self._max_items = max_items
        self._buffer: List[T] = []
        self._cursor: Optional[str] = None
        self._started = False
        self._exhausted = False
        self._yielded = 0

    def __aiter__(self) -> AsyncIterator[T]:
        return self

    async def __anext__(self) -> T:
        if self._max_items is not None and self._yielded >= self._max_items:
            raise StopAsyncIteration
        if not self._buffer and not self._exhausted:
            await self._fill()
        if not self._buffer:
            raise StopAsyncIteration
        item = self._buffer.pop(0)
        self._yielded += 1
        return item

    async def _fill(self) -> None:
        if self._started and self._cursor is None:
            self._exhausted = True
            return
        items, next_cursor = await self._fetcher(self._cursor)
        self._started = True
        self._buffer.extend(items)
        self._cursor = next_cursor
        if not items or next_cursor is None:
            self._exhausted = True


def extract_items(payload: Any, key: str = "factors") -> List[Dict[str, Any]]:
    """Pull ``key`` out of a response payload regardless of wrapper shape."""
    if isinstance(payload, dict):
        inner = payload.get(key)
        if isinstance(inner, list):
            return inner
    if isinstance(payload, list):
        return payload
    return []


__all__ = [
    "PageInfo",
    "OffsetPaginator",
    "CursorPaginator",
    "AsyncOffsetPaginator",
    "AsyncCursorPaginator",
    "extract_items",
]

File: greenlang/factors/sdk/ts/src/pagination.ts
/**
 * Async pagination helpers, parity with the Python sync/async paginators.
 *
 * Both `OffsetPaginator` and `CursorPaginator` implement the async
 * iterator protocol so callers can simply `for await ... of` them.
 */

export interface PageInfo {
  page: number;
  pageSize: number;
  totalCount: number | null;
  nextCursor: string | null;
  itemsYielded: number;
}

export type OffsetFetcher<T> = (
  offset: number,
  limit: number,
) => Promise<{ items: T[]; totalCount: number | null }>;

export type CursorFetcher<T> = (
  cursor: string | null,
) => Promise<{ items: T[]; nextCursor: string | null }>;

/**
 * Async offset/limit iterator. Fetcher receives `(offset, limit)` and
 * returns `{ items, totalCount }`. Iteration stops when:
 *   - `items` is empty, OR
 *   - offset + yielded >= totalCount (if known), OR
 *   - `maxItems` has been reached.
 */
export class OffsetPaginator<T> implements AsyncIterable<T>, AsyncIterator<T> {
  private readonly fetcher: OffsetFetcher<T>;
  private readonly pageSize: number;
  private readonly maxItems: number | undefined;
  private offset: number;
  private buffer: T[] = [];
  private totalCount: number | null = null;
  private exhausted = false;
  private yielded = 0;

  constructor(
    fetcher: OffsetFetcher<T>,
    opts: {
      pageSize?: number;
      startOffset?: number;
      maxItems?: number;
    } = {},
  ) {
    this.fetcher = fetcher;
    this.pageSize = Math.max(1, opts.pageSize ?? 100);
    this.offset = opts.startOffset ?? 0;
    this.maxItems = opts.maxItems;
  }

  [Symbol.asyncIterator](): AsyncIterator<T> {
    return this;
  }

  async next(): Promise<IteratorResult<T>> {
    if (this.maxItems !== undefined && this.yielded >= this.maxItems) {
      return { done: true, value: undefined };
    }
    if (this.buffer.length === 0 && !this.exhausted) {
      await this.fill();
    }
    if (this.buffer.length === 0) {
      return { done: true, value: undefined };
    }
    const item = this.buffer.shift() as T;
    this.yielded += 1;
    return { done: false, value: item };
  }

  get total(): number | null {
    return this.totalCount;
  }

  /** Collect *all* remaining items into an array. */
  async toArray(): Promise<T[]> {
    const out: T[] = [];
    for (;;) {
      const r = await this.next();
      if (r.done) break;
      out.push(r.value);
    }
    return out;
  }

  private async fill(): Promise<void> {
    let limit = this.pageSize;
    if (this.maxItems !== undefined) {
      limit = Math.min(limit, this.maxItems - this.yielded);
      if (limit <= 0) {
        this.exhausted = true;
        return;
      }
    }
    const { items, totalCount } = await this.fetcher(this.offset, limit);
    this.buffer.push(...items);
    this.offset += items.length;
    if (totalCount !== null && totalCount !== undefined) {
      this.totalCount = totalCount;
    }
    if (
      items.length === 0 ||
      (this.totalCount !== null && this.offset >= this.totalCount)
    ) {
      this.exhausted = true;
    }
  }
}

/**
 * Async cursor iterator. Fetcher receives `cursor` (null on the first
 * call) and returns `{ items, nextCursor }`. Iteration stops when
 * `nextCursor` is null or `maxItems` is reached.
 */
export class CursorPaginator<T> implements AsyncIterable<T>, AsyncIterator<T> {
  private readonly fetcher: CursorFetcher<T>;
  private readonly maxItems: number | undefined;
  private buffer: T[] = [];
  private cursor: string | null = null;
  private started = false;
  private exhausted = false;
  private yielded = 0;

  constructor(fetcher: CursorFetcher<T>, opts: { maxItems?: number } = {}) {
    this.fetcher = fetcher;
    this.maxItems = opts.maxItems;
  }

  [Symbol.asyncIterator](): AsyncIterator<T> {
    return this;
  }

  async next(): Promise<IteratorResult<T>> {
    if (this.maxItems !== undefined && this.yielded >= this.maxItems) {
      return { done: true, value: undefined };
    }
    if (this.buffer.length === 0 && !this.exhausted) {
      await this.fill();
    }
    if (this.buffer.length === 0) {
      return { done: true, value: undefined };
    }
    const item = this.buffer.shift() as T;
    this.yielded += 1;
    return { done: false, value: item };
  }

  get nextCursor(): string | null {
    return this.cursor;
  }

  async toArray(): Promise<T[]> {
    const out: T[] = [];
    for (;;) {
      const r = await this.next();
      if (r.done) break;
      out.push(r.value);
    }
    return out;
  }

  private async fill(): Promise<void> {
    if (this.started && this.cursor === null) {
      this.exhausted = true;
      return;
    }
    const { items, nextCursor } = await this.fetcher(this.cursor);
    this.started = true;
    this.buffer.push(...items);
    this.cursor = nextCursor;
    if (items.length === 0 || nextCursor === null) {
      this.exhausted = true;
    }
  }
}

/** Pull `key` out of a response payload regardless of wrapper shape. */
export function extractItems<T = unknown>(
  payload: unknown,
  key = 'factors',
): T[] {
  if (payload && typeof payload === 'object' && !Array.isArray(payload)) {
    const inner = (payload as Record<string, unknown>)[key];
    if (Array.isArray(inner)) return inner as T[];
  }
  if (Array.isArray(payload)) return payload as T[];
  return [];
}
